hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/AbstractS3GuardDiagnostic.java

```diff
@@ -0,0 +1,221 @@
+   * @param filesystem filesystem
+   * @param store optional metastore.
+   * @param uri URI. Must be set if filesystem == null.
+   */
+  public AbstractS3GuardDiagnostic(
+      final String name,
+      @Nullable final S3AFileSystem filesystem,
+      @Nullable final DynamoDBMetadataStore store,
+      @Nullable final URI uri) {
+    super(name);
+    this.store = store;
+    this.filesystem = filesystem;
+    if (store == null) {
+      require(filesystem != null, "No filesystem or URI");
+      bindStore(filesystem);
+    }
+    if (uri == null) {
+      require(filesystem != null, "No filesystem or URI");
+      setUri(filesystem.getUri());
+    } else {
+      setUri(uri);
+    }
+  }
+
+  /**
+   * Require a condition to hold, otherwise an exception is thrown.
+   * @param condition condition to be true
+   * @param error text on failure.
+   * @throws ServiceLaunchException if the condition is not met
+   */
+  protected static void require(boolean condition, String error) {
+    if (!condition) {
+      throw failure(error);
+    }
+  }
+
+  /**
+   * Generate a failure exception for throwing.
+   * @param message message
+   * @param ex optional nested exception.
+   * @return an exception to throw
+   */
+  protected static ServiceLaunchException failure(String message, Throwable ex) {
+    return new ServiceLaunchException(LauncherExitCodes.EXIT_FAIL, message, ex);
+  }
+
+  /**
+   * Generate a failure exception for throwing.
+   * @param message message
+   * @return an exception to throw
+   */
+  protected static ServiceLaunchException failure(String message) {
+    return new ServiceLaunchException(LauncherExitCodes.EXIT_FAIL, message);
+  }
+
+  @Override
+  public Configuration bindArgs(final Configuration config,
+      final List<String> args)
+      throws Exception {
+    this.arguments = args;
+    return super.bindArgs(config, args);
+  }
+
+  /**
+   * Get the argument list.
+   * @return the argument list.
+   */
+  protected List<String> getArguments() {
+    return arguments;
+  }
+
+  /**
+   * Bind to the store from a CLI argument.
+   * @param fsURI filesystem URI
+   * @throws IOException failure
+   */
+  protected void bindFromCLI(String fsURI)
+      throws IOException {
+    Configuration conf = getConfig();
+    setUri(fsURI);
+    FileSystem fs = FileSystem.get(getUri(), conf);
+    require(fs instanceof S3AFileSystem,
+        "Not an S3A Filesystem:  " + fsURI);
+    filesystem = (S3AFileSystem) fs;
+    bindStore(filesystem);
+    setUri(fs.getUri());
+  }
+
+  /**
+   * Binds the {@link #store} field to the metastore of
+   * the filesystem -which must have a DDB metastore.
+   * @param fs filesystem to bind the store to.
+   */
+  private void bindStore(final S3AFileSystem fs) {
+    require(fs.hasMetadataStore(),
+        "Filesystem has no metadata store: " + fs.getUri());
+    MetadataStore ms = fs.getMetadataStore();
+    require(ms instanceof DynamoDBMetadataStore,
```

Comment:
Why do we bind to dynamodbms? Why only dynamo? Based on the last comments on the Localms removal (https://issues.apache.org/jira/browse/HADOOP-16326) it's good to have local to find out errors. I don't support this. Or name this test to dynamo, as this is dynamo specific.

Reply:
we're grabbing the DDB table and doing things with it -those scan commands in particular

Reply:
Maybe we should do some diagnostics on localms as well in the future.